### PR TITLE
[openwebnet] [wip] Add Auxiliary(AUX) message support to the binding (WHO=9)

### PR DESCRIPTION
**Pull request by fabgio**
[openwebnet] This PR adds support for Auxiliary (WHO=9) to the [openwebnet] binding.
With this addition BTicino BUS AUX (WHO=9) can be detected and activated from OH3, enabling OH3 to detect AUX switches allowing, for example to control alarm systems of a BTicino installation.
Configuration for new things and channel has been added to the add-on README.
Waiting for testing of this new addition on a real plant

### CODE REVIEW

**Comment by marcelrv** [comment]
suggest to rebase your PR so it does not pull all these other commits

**Comment by fabgio** [comment]
Rebasing commits seems  to be a dangerous operation as  I am afraid of loosing commits, could I skip this task? otherwise I need help

**Comment by fabgio** [comment]
As you suggested, I rebased commits as much as possible

**Comment by hmerk** [comment]
Still not ok, as you can see with all required reviews.

**Comment by fabgio** [comment]
Ok i made some interactive rebasing so I cleaned  up the PR.  I apologize for the inconvenence; this my fist PR!

**Comment by fabgio** [comment]
Any news?

**Comment by wborn** [comment]
Thanks for the PR! Perhaps @mvalla can help with reviewing these changes?

**Comment by mvalla** [comment]
> Perhaps @mvalla can help with reviewing these changes?

Sure, I will do it, and test the binding as well.

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAuxiliaryHandler.java [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+    /**
+     * Handles Auxiliary switch command for a channel
+     *
+     * @param channel the channel
+     * @param command the Command
+     */
+    @Override
+    protected void handleChannelCommand(ChannelUID channel, Command command) {
+        logger.debug("handleSwitchCommand() (command={} - channel={})", command, channel);
```
Comment (mvalla) [inline bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAuxiliaryHandler.java:69]:
```suggestion
        logger.debug("AUX handleChannelCommand() (command={} - channel={})", command, channel);
```

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAuxiliaryHandler.java [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+            if (command instanceof OnOffType) {
+                try {
+                    if (OnOffType.ON.equals(command)) {
+                        send(Auxiliary.requestTurnOn(toWhere(channel.getId())));
+                    } else if (OnOffType.OFF.equals(command)) {
+                        send(Auxiliary.requestTurnOff(toWhere(channel.getId())));
+                    }
+                } catch (OWNException e) {
+                    logger.warn("Exception while processing command {}: {}", command, e.getMessage());
+                }
```
Comment (mvalla) [inline bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAuxiliaryHandler.java:80]: instead of calling toWhere (which is useless here), just check deviceWhere is not null like is done in requestStatus() before calling send()

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAuxiliaryHandler.java [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+        logger.debug("requestChannelState() thingUID={} channel={}", thing.getUID(), channel.getId());
+        requestStatus(channel.getId());
+    }
+
+    /** helper method to request auxiliary status based on channel */
+    private void requestStatus(String channelId) {
+        Where w = deviceWhere;
+        if (w != null) {
+            try {
+                Response res = send(Auxiliary.requestStatus(toWhere(channelId)));
```
Comment (mvalla) [inline bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAuxiliaryHandler.java:98]:
```suggestion
                Response res = send(Auxiliary.requestStatus(w.value()));
```

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAuxiliaryHandler.java [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+        OpenWebNetBridgeHandler brH = bridgeHandler;
+        if (brH != null) {
+            if (brH.isBusGateway() && refreshAll) {
+                long now = System.currentTimeMillis();
+                if (now - lastAllDevicesRefreshTS > ALL_DEVICES_REFRESH_INTERVAL_MSEC) {
+                    try {
+                        send(Auxiliary.requestStatus(WhereAuxiliary.GENERAL.value()));
+                        lastAllDevicesRefreshTS = now;
+                    } catch (OWNException e) {
+                        logger.warn("Exception while requesting all devices refresh: {}", e.getMessage());
```
Comment (mvalla) [inline bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAuxiliaryHandler.java:124]:
```suggestion
                        logger.warn("Exception while requesting all AUX devices refresh: {}", e.getMessage());
```

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAuxiliaryHandler.java [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+                    }
+                } else {
+                    logger.debug("Refresh all devices just sent...");
```
Comment (mvalla) [inline bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAuxiliaryHandler.java:127]:
```suggestion
                    logger.debug("Refresh all AUX devices just sent...");
```

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAuxiliaryHandler.java [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+        if (w != null) {
+            OpenWebNetBridgeHandler brH = bridgeHandler;
+            if (brH != null) {
+                return w.value();
+            } else if (channelId.equals(CHANNEL_SWITCH)) {
+                return w.value();
+            }
+        }
+        return null;
+    }
```
Comment (mvalla) [inline bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAuxiliaryHandler.java:159]: This method like it's written does not anything useful. It's better to remove it and just check deviceWhere is not null when it is used (like in requestStatus()).

File: bundles/org.openhab.binding.openwebnet/README.md [hunk 7]
```diff
@@ -44,8 +44,10 @@ The following Things and OpenWebNet `WHOs` are supported:
...
 | Temperature Control  | `4`          | `bus_thermo_zone`, `bus_thermo_sensor` | Thermo zones management and temperature sensors (probes). NOTE Central Units (4 or 99 zones) are not fully supported yet. See [Channels - Thermo](#configuring-thermo) for more details. | Successfully tested: H/LN4691, HS4692, KG4691; thermo sensors: L/N/NT4577 + 3455 |
+| Auxiliary support    | `9`          | `aux_on_off_switch`                    | AUX switches                                                  | Not tested yet 
 | CEN & CEN+ Scenarios  | `15` & `25`  | `bus_cen_scenario_control`, `bus_cenplus_scenario_control` | CEN/CEN+ scenarios events and virtual activation | Successfully tested: scenario buttons: HC/HD/HS/L/N/NT4680 |
-| Energy Management    | `18`         | `bus_energy_meter`                  | Energy Management                                                | Successfully tested: F520, F521 |
+| Energy Management    | `18`         | `bus_energy_meter`                  | Energy Management| Successfully tested: F520, F521 |
+|
```
Comment (mvalla) [inline bundles/org.openhab.binding.openwebnet/README.md:50]: Why this change? Also what is the purpose of the last "|" ?

File: bundles/org.openhab.binding.openwebnet/README.md [hunk 8]
```diff
@@ -44,8 +44,10 @@ The following Things and OpenWebNet `WHOs` are supported:
 | Lighting             | `1`          | `bus_on_off_switch`, `bus_dimmer`   | BUS switches and dimmers                                         | Successfully tested: F411/2, F411/4, F411U2, F422, F429. Some discovery issues reported with F429 (DALI Dimmers)  |
 | Automation           | `2`          | `bus_automation`                    | BUS roller shutters, with position feedback and auto-calibration | Successfully tested: LN4672M2  |
 | Temperature Control  | `4`          | `bus_thermo_zone`, `bus_thermo_sensor` | Thermo zones management and temperature sensors (probes). NOTE Central Units (4 or 99 zones) are not fully supported yet. See [Channels - Thermo](#configuring-thermo) for more details. | Successfully tested: H/LN4691, HS4692, KG4691; thermo sensors: L/N/NT4577 + 3455 |
+| Auxiliary support    | `9`          | `aux_on_off_switch`                    | AUX switches                                                  | Not tested yet
```
Comment (mvalla) [inline bundles/org.openhab.binding.openwebnet/README.md:47]: "Not tested yet" : if no real device has been tested, just leave blank

**Review by mvalla — CHANGES_REQUESTED** [review]
Please check  review suggestions.  After they are fixed I will do a quick test of the binding.

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAuxiliaryHandler.java (hunk 1 above)
Comment (fabgio) [reply bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAuxiliaryHandler.java:69]: Fixed

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAuxiliaryHandler.java (hunk 3 above)
Comment (fabgio) [reply bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAuxiliaryHandler.java:98]: ok

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAuxiliaryHandler.java (hunk 4 above)
Comment (fabgio) [reply bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAuxiliaryHandler.java:124]: fixed

File: bundles/org.openhab.binding.openwebnet/README.md (hunk 7 above)
Comment (fabgio) [reply bundles/org.openhab.binding.openwebnet/README.md:50]: fixed

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAuxiliaryHandler.java (hunk 6 above)
Comment (fabgio) [reply bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAuxiliaryHandler.java:159]: fixed

**Comment by mvalla** [comment]
the following additional changes are needed:
- new Thing `bus_aux` xml definition, and new channel-type `auxiliary` definition
- README:  Things and Channels tables should be updated, examples (.things, .items, .sitemap) are missing for AUX
- AUX Thing discovery should be added to openwebnet4j BUSGateway.java (see also class  `OpenWebNetDeviceDiscoveryService.java`)
- rebase on current main
